src/libsystemd/sd-bus/bus-socket.c

```diff
@@ -607,39 +622,42 @@ static void bus_get_peercred(sd_bus *b) {
 }
 
 static int bus_socket_start_auth_client(sd_bus *b) {
-        size_t l;
-        const char *auth_suffix, *auth_prefix;
+        static char sasl_auth_anonymous[] = {
+                /*
+                 * We use an arbitrary trace-string for the ANONYMOUS authentication. It can be used by the
+                 * message broker to aid debugging of clients. We fully anonymize the connection and use a
+                 * static default.
+                 */
+                "\0AUTH ANONYMOUS\r\n"
+                /* HEX a n o n y m o u s */
+                "DATA 616e6f6e796d6f7573\r\n"
+        };
+        static char sasl_auth_external[] = {
+                "\0AUTH EXTERNAL\r\n"
+                "DATA\r\n"
+        };
+        static char sasl_negotiate_unix_fd[] = {
+                "NEGOTIATE_UNIX_FD\r\n"
+        };
+        static char sasl_begin[] = {
+                "BEGIN\r\n"
+        };
 
         assert(b);
 
-        if (b->anonymous_auth) {
-                auth_prefix = "\0AUTH ANONYMOUS ";
-
-                /* For ANONYMOUS auth we send some arbitrary "trace" string */
-                l = 9;
-                b->auth_buffer = hexmem("anonymous", l);
-        } else {
-                char text[DECIMAL_STR_MAX(uid_t) + 1];
-
-                auth_prefix = "\0AUTH EXTERNAL ";
-
-                xsprintf(text, UID_FMT, geteuid());
-
-                l = strlen(text);
-                b->auth_buffer = hexmem(text, l);
-        }
-
-        if (!b->auth_buffer)
-                return -ENOMEM;
+        if (b->anonymous_auth)
+                b->auth_iovec[0] = IOVEC_MAKE(sasl_auth_anonymous,
+                                              1 + strlen(sasl_auth_anonymous + 1));
+        else
+                b->auth_iovec[0] = IOVEC_MAKE(sasl_auth_external,
+                                              1 + strlen(sasl_auth_external + 1));
 
         if (b->accept_fd)
-                auth_suffix = "\r\nNEGOTIATE_UNIX_FD\r\nBEGIN\r\n";
+                b->auth_iovec[1] = IOVEC_MAKE_STRING(sasl_negotiate_unix_fd);
         else
-                auth_suffix = "\r\nBEGIN\r\n";
+                b->auth_iovec[1] = IOVEC_MAKE(NULL, 0);
```

Comment:
doesn't matter much, but I'd prefer if we'd have b->n_auth_iovec as a counter that specifies exactly how many entries the iovec array has so that we set it to either 2 or 3 instead of adding an empty iovec into the middle

Reply:
Sure.